src/Control/Effect/Cut.hs

```diff
@@ -48,53 +49,64 @@ call m = send (Call m ret)
 
 -- | Commit to the current branch, preventing backtracking within the nearest enclosing 'call' (if any) on failure.
 --
---   prop> run (runNonDet (runCut (cut *> pure a <|> pure b))) == [a, b]
+--   prop> run (runNonDet (runCut (pure a <|> cut *> pure b))) == [a, b]
+--   prop> run (runNonDet (runCut (cut *> pure a <|> pure b))) == [a]
```

Comment:
Importantly, `cut` now prunes when successful as well. This is the behaviour of `cut` in Hinze’s _Deriving Backtracking Monad Transformers_, and I believe it’s correct w.r.t. our desired semantics: e.g. in a parser, you want to avoid backtracking after a successful `cut` as well as a failed one in order to limit ambiguity.

Note that in Kiselyov et al’s _Backtracking, Interleaving, and Terminating Monad Transformers_, they discuss Prolog-style _cut_’s conflation of negation with pruning; this criticism applies to `Cut` as well. However, it’s reasonable to consider adding these as orthogonal features using a new effect, and in the meantime, `cut` is a reasonably well-understood logical operator.